[Build 40](https://ci.wildfly.org/viewLog.html?buildId=124787&buildTypeId=Xnio_33_LinuxIbmJdk8) outcome was **FAILURE** using a merge of 94ef04e0d5689c2eeb841759087de8f58a26ffbf
Summary: Tests failed: 7 (1 new), passed: 945, ignored: 18 Build time: 00:01:19
### Failed tests
```
org.xnio.nio.test.MultiThreadedNioSslTcpConnectionTestCase.twoWayTransfer: java.lang.AssertionError
	at org.xnio.nio.test.AbstractNioSslTcpTest$7.run(AbstractNioSslTcpTest.java:240)
	at org.xnio.nio.test.AbstractNioTcpTest.doConnectionTest(AbstractNioTcpTest.java:102)
	at org.xnio.nio.test.NioSslTcpConnectionTestCase.doConnectionTest(NioSslTcpConnectionTestCase.java:115)
	at org.xnio.nio.test.AbstractNioSslTcpTest.twoWayTransfer(AbstractNioSslTcpTest.java:237)
------- Stderr: -------
09:35:33,530 INFO  (main) [TEST] <AbstractNioSslTcpTest.java:231> Test: twoWayTransfer
09:35:33,534 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:336> server handle readable
09:35:33,539 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,540 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,541 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,541 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,541 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,542 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,542 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,542 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,543 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,543 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:355> server handle writable
09:35:33,549 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:255> client handle readable
09:35:33,557 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,557 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,558 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,558 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,558 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,559 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,559 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,559 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,560 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,560 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,560 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,560 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,561 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,561 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,561 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,562 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,562 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,562 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,562 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,563 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,563 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,563 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,564 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,564 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,564 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,565 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,565 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,565 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,565 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,566 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,566 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
09:35:33,566 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable


org.xnio.racecondition.ResumeReadsOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.ResumeWritesOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetReadListenerOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetReadReadyOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetWriteListenerOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetWriteReadyOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)
------- Stdout: -------
byteman jar is /store/repository/org/jboss/byteman/byteman/3.0.3/byteman-3.0.3.jar
Setting org.jboss.byteman.allow.config.update=true
------- Stderr: -------
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:408)
	at sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(InstrumentationImpl.java:433)
	at com.ibm.tools.attach.javaSE.Attachment.loadAgentLibraryImpl(Native Method)
	at com.ibm.tools.attach.javaSE.Attachment.loadAgentLibrary(Attachment.java:309)
	at com.ibm.tools.attach.javaSE.Attachment.parseLoadAgent(Attachment.java:287)
	at com.ibm.tools.attach.javaSE.Attachment.doCommand(Attachment.java:168)
	at com.ibm.tools.attach.javaSE.Attachment.run(Attachment.java:128)
Caused by: java.lang.UnsupportedOperationException: adding retransformable transformers is not supported in this environment
	at sun.instrument.InstrumentationImpl.addTransformer(InstrumentationImpl.java:100)
	at org.jboss.byteman.agent.Main.premain(Main.java:275)
	at org.jboss.byteman.agent.Main.agentmain(Main.java:305)
	... 11 more


```

--- a/api/src/main/java/org/xnio/XnioWorker.java
+++ b/api/src/main/java/org/xnio/XnioWorker.java
@@ -131,18 +131,20 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
             if (executorService instanceof EnhancedQueueExecutor) {
                 taskPool = new ExternalTaskPool(
                         new EnhancedQueueExecutorTaskPool((EnhancedQueueExecutor) executorService));
+            } else if (executorService instanceof ThreadPoolExecutor) {
+                taskPool = new ExternalTaskPool(new ThreadPoolExecutorTaskPool((ThreadPoolExecutor) executorService));
             } else {
                 taskPool = new ExternalTaskPool(new ExecutorServiceTaskPool(executorService));
             }
         } else if (EnhancedQueueExecutor.DISABLE_HINT) {
             final int poolSize = max(builder.getMaxWorkerPoolSize(), builder.getCoreWorkerPoolSize());
-            taskPool = new ThreadPoolExecutorTaskPool(
+            taskPool = new ThreadPoolExecutorTaskPool(new DefaultThreadPoolExecutor(
                 poolSize,
                 poolSize,
                 builder.getWorkerKeepAlive(), TimeUnit.MILLISECONDS,
                 new LinkedBlockingDeque<>(),
                 new WorkerThreadFactory(builder.getThreadGroup(), builder.getWorkerStackSize(), markThreadAsDaemon),
-                terminationTask);
+                terminationTask));
         } else {
             taskPool = new EnhancedQueueExecutorTaskPool(new EnhancedQueueExecutor.Builder()
                 .setCorePoolSize(builder.getCoreWorkerPoolSize())
@@ -1298,10 +1300,10 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
         int getQueueSize();
     }
 
-    static class ThreadPoolExecutorTaskPool extends ThreadPoolExecutor implements TaskPool {
+    static class DefaultThreadPoolExecutor extends ThreadPoolExecutor {
         private final Runnable terminationTask;
 
-        ThreadPoolExecutorTaskPool(final int corePoolSize, final int maximumPoolSize, final long keepAliveTime, final TimeUnit unit, final BlockingQueue<Runnable> workQueue, final ThreadFactory threadFactory, final Runnable terminationTask) {
+        DefaultThreadPoolExecutor(final int corePoolSize, final int maximumPoolSize, final long keepAliveTime, final TimeUnit unit, final BlockingQueue<Runnable> workQueue, final ThreadFactory threadFactory, final Runnable terminationTask) {
             super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue, threadFactory);
             this.terminationTask = terminationTask;
         }
@@ -1323,9 +1325,68 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
                 super.setMaximumPoolSize(size);
             }
         }
+    }
 
+    static class ThreadPoolExecutorTaskPool implements TaskPool {
+        private final ThreadPoolExecutor delegate;
+
+        ThreadPoolExecutorTaskPool(final ThreadPoolExecutor delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void shutdown() {
+            delegate.shutdown();
+        }
+
+        @Override
+        public List<Runnable> shutdownNow() {
+            return delegate.shutdownNow();
+        }
+
+        @Override
+        public void execute(final Runnable command) {
+            delegate.execute(command);
+        }
+
+        @Override
+        public int getCorePoolSize() {
+            return delegate.getCorePoolSize();
+        }
+
+        @Override
+        public int getMaximumPoolSize() {
+            return delegate.getMaximumPoolSize();
+        }
+
+        @Override
+        public long getKeepAliveTime(final TimeUnit unit) {
+            return delegate.getKeepAliveTime(unit);
+        }
+
+        @Override
+        public void setCorePoolSize(final int size) {
+            delegate.setCorePoolSize(size);
+        }
+
+        @Override
+        public void setMaximumPoolSize(final int size) {
+            delegate.setMaximumPoolSize(size);
+        }
+
+        @Override
+        public void setKeepAliveTime(final long time, final TimeUnit unit) {
+            delegate.setKeepAliveTime(time, unit);
+        }
+
+        @Override
+        public int getActiveCount() {
+            return delegate.getActiveCount();
+        }
+
+        @Override
         public int getQueueSize() {
-            return getQueue().size();
+            return delegate.getQueue().size();
         }
     }
 
@@ -1430,9 +1491,6 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
         }
     }
 
-    /**
-     * {@link ExternalTaskPool} wrapper to prevent an delegate {@link TaskPool} from being shut down by the worker.
-     */
     static class ExternalTaskPool implements TaskPool {
         private final TaskPool delegate;
 

--- a/api/src/main/java/org/xnio/XnioWorker.java
+++ b/api/src/main/java/org/xnio/XnioWorker.java
@@ -861,7 +861,7 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
             .create();
 
     public boolean supportsOption(final Option<?> option) {
-        return taskPool instanceof ExecutorServiceTaskPool ? EXTERNAL_POOL_OPTIONS.contains(option) : OPTIONS.contains(option);
+        return taskPool instanceof ExternalTaskPool ? EXTERNAL_POOL_OPTIONS.contains(option) : OPTIONS.contains(option);
     }
 
     public <T> T getOption(final Option<T> option) throws IOException {

--- a/api/src/main/java/org/xnio/XnioWorker.java
+++ b/api/src/main/java/org/xnio/XnioWorker.java
@@ -129,9 +129,10 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
         final ExecutorService executorService = builder.getExternalExecutorService();
         if (executorService != null) {
             if (executorService instanceof EnhancedQueueExecutor) {
-                taskPool = new EnhancedQueueExecutorTaskPool((EnhancedQueueExecutor) executorService);
+                taskPool = new ExternalTaskPool(
+                        new EnhancedQueueExecutorTaskPool((EnhancedQueueExecutor) executorService));
             } else {
-                taskPool = new ExternalTaskPool(executorService);
+                taskPool = new ExternalTaskPool(new ExecutorServiceTaskPool(executorService));
             }
         } else if (EnhancedQueueExecutor.DISABLE_HINT) {
             final int poolSize = max(builder.getMaxWorkerPoolSize(), builder.getCoreWorkerPoolSize());
@@ -860,7 +861,7 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
             .create();
 
     public boolean supportsOption(final Option<?> option) {
-        return taskPool instanceof ExternalTaskPool ? EXTERNAL_POOL_OPTIONS.contains(option) : OPTIONS.contains(option);
+        return taskPool instanceof ExecutorServiceTaskPool ? EXTERNAL_POOL_OPTIONS.contains(option) : OPTIONS.contains(option);
     }
 
     public <T> T getOption(final Option<T> option) throws IOException {
@@ -1380,19 +1381,19 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
         }
     }
 
-    static class ExternalTaskPool implements TaskPool {
+    static class ExecutorServiceTaskPool implements TaskPool {
         private final ExecutorService delegate;
 
-        ExternalTaskPool(final ExecutorService delegate) {
+        ExecutorServiceTaskPool(final ExecutorService delegate) {
             this.delegate = delegate;
         }
 
         public void shutdown() {
-            // no operation
+            delegate.shutdown();
         }
 
         public List<Runnable> shutdownNow() {
-            return Collections.emptyList();
+            return delegate.shutdownNow();
         }
 
         public void execute(final Runnable command) {
@@ -1426,6 +1427,72 @@ public abstract class XnioWorker extends AbstractExecutorService implements Conf
 
         public int getQueueSize() {
             return -1;
+        }
+    }
+
+    /**
+     * {@link ExternalTaskPool} wrapper to prevent an delegate {@link TaskPool} from being shut down by the worker.
+     */
+    static class ExternalTaskPool implements TaskPool {
+        private final TaskPool delegate;
+
+        ExternalTaskPool(final TaskPool delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void shutdown() {
+            // no operation
+        }
+
+        @Override
+        public List<Runnable> shutdownNow() {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public void execute(final Runnable command) {
+            delegate.execute(command);
+        }
+
+        @Override
+        public int getCorePoolSize() {
+            return delegate.getCorePoolSize();
+        }
+
+        @Override
+        public int getMaximumPoolSize() {
+            return delegate.getMaximumPoolSize();
+        }
+
+        @Override
+        public long getKeepAliveTime(final TimeUnit unit) {
+            return delegate.getKeepAliveTime(unit);
+        }
+
+        @Override
+        public void setCorePoolSize(final int size) {
+            delegate.setCorePoolSize(size);
+        }
+
+        @Override
+        public void setMaximumPoolSize(final int size) {
+            delegate.setMaximumPoolSize(size);
+        }
+
+        @Override
+        public void setKeepAliveTime(final long time, final TimeUnit unit) {
+            delegate.setKeepAliveTime(time, unit);
+        }
+
+        @Override
+        public int getActiveCount() {
+            return delegate.getActiveCount();
+        }
+
+        @Override
+        public int getQueueSize() {
+            return delegate.getQueueSize();
         }
     }
 }

